File: r/day15_part1_2016.R

discs <- readLines("input.txt")
total_positions <- as.integer(sub(".*has (\\d+) positions.*", "\\1", discs))
start_position <- as.integer(sub(".*position (\\d+).*", "\\1", discs))

time <- 0
while (TRUE) {
  if (sum((start_position + time + seq_along(total_positions)) %% total_positions) == 0) {
    cat(time, "\n")
    break
  }
  time <- time + 1
}

File: matlab/day15_part1_2016.m

function main
    fid = fopen('input.txt', 'r');
    discs = [];
    while ~feof(fid)
        line = fgetl(fid);
        [~, ~, ~, ~, matches] = regexp(line, 'Disc #\d+ has (\d+) positions; at time=0, it is at position (\d+).');
        total_positions = str2double(matches{1}{1});
        start_position = str2double(matches{1}{2});
        discs = [discs; struct('total_positions', total_positions, 'start_position', start_position)];
    end
    fclose(fid);

    time = 0;
    while true
        valid = true;
        for i = 1:length(discs)
            if mod(discs(i).start_position + time + i, discs(i).total_positions) ~= 0
                valid = false;
                break;
            end
        end
        if valid
            disp(time);
            break;
        end
        time = time + 1;
    end
end
